keps/sig-node/xxxx-container-oom-kill-mode/README.md

```diff
@@ -0,0 +1,729 @@
+- Reports of potential issues affecting customers due to the behavioral change
+- Multi-process workloads experienced complete container restarts when only a single process needed to be terminated[^2]
+
+These reports led to [PR #122813](https://github.com/kubernetes/kubernetes/pull/122813), which attempted to add a kubelet flag but was closed after community discussion concluded that container-level configuration was the proper solution[^3]. The consensus was that node-level configuration cannot adequately address the needs of heterogeneous workloads.
+
+This KEP also deprecates the `singleProcessOOMKill` kubelet flag for removal in v1.38 (GA). Container-level configuration provides better granularity and eliminates the complexity of maintaining both node-level and container-level settings.
+
+### Goals
+
+- Add a per-container `oomKillMode` field to allow container-level OOM behavior configuration
+- Maintain full backward compatibility with the existing `singleProcessOOMKill` kubelet configuration during migration
+- Deprecate and remove `singleProcessOOMKill` kubelet flag by v1.38 (GA) to simplify configuration
+
+### Non-Goals
+
+- Provide cgroups v1 and Windows support
+
+## Proposal
+
+Add an `oomKillMode` field to the Container specification in the core v1 API. This design choice is deliberate and based on several key considerations:
+
+### Why Container-level Configuration?
+
+1. **Granular Control**: Different containers in the same pod may have different OOM requirements:
+
+   - A sidecar container might benefit from single-process kills
+   - The main application container might require group kills for consistency
+
+2. **Natural Alignment**: OOM behavior is inherently a container-level concern:
+
+   - Memory limits are set per container
+   - OOMScoreAdj is calculated per container based on QoS
+   - cgroups are managed at the container level
+
+3. **Consistency with Existing APIs**: Follows the pattern of other container-level resource configurations:
+   - Resources (limits/requests) are container-level
+   - SecurityContext can be set at both pod and container level
+   - The existing `singleProcessOOMKill` implementation operates per container
+
+### API Design
+
+```go
+type Container struct {
+    // ... existing fields ...
+
+    // Compute Resources required by this container.
+    // Cannot be updated.
+    // More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
+    // +optional
+    Resources ResourceRequirements `json:"resources,omitempty" protobuf:"bytes,8,opt,name=resources"`
+
+    // OOMKillMode specifies how the OOM killer behaves for this container.
+    // - "Single": only the process that triggered OOM is killed
+    // - "Group": all processes in the container are killed (cgroup v2 default)
+    // If not specified, the behavior is determined by the kubelet configuration.
```

Comment:
If `OOMKillMode` is not specified, the kubelet will determine the default value. It might be helpful to write the determined value into the ContainerStatus so that users can check it.
https://github.com/kubernetes/kubernetes/blob/v1.33.4/staging/src/k8s.io/api/core/v1/types.go#L3152